### PR TITLE
Import-Export Improvement

### PR DESCRIPTION
**Pull request by jat-odoo**
Description of the issue/feature this PR addresses:
* Task: https://www.odoo.com/web#id=40692&action=327&model=project.task&view_type=form&menu_id=4720
* Pad: https://pad.odoo.com/p/r.cd3cc8461f8fee493394916c724fb49a

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product/static/src/js/product_template_import.js [hunk 1]
```diff
@@ -1,19 +1,75 @@
...
+
+    /**
+     * Set template link based on some dynamic cases:
+     * - If pricelist not activated or pricelist activated with formulas
+     * - If default option of pricelist is activated (several prices on products)
+     * - If purchase module then set different template for products
+     *
+     * @override
+     */
+    renderImportLink: function () {
```
Comment (xmo-odoo) [inline addons/product/static/src/js/product_template_import.js]: The entire selection of import templates should be moved to the server side and properly overloaded on a per-model basis. The JS side should live only in the base import module and just call out over RPC to get the relevant link & label.

File: addons/base_import/models/base_import.py [hunk 2]
```diff
@@ -497,6 +514,7 @@ def parse_preview(self, options, count=10):
                 'headers_type': header_types or False,
                 'preview': preview,
                 'options': options,
+                'advanced_mode': options.get('advanced') if options.get('keep_matches') else any([len(models.fix_import_export_id_paths(col)) > 1 for col in headers or []]),
```
Comment (xmo-odoo) [inline addons/base_import/models/base_import.py]: What's the difference between advanced & keep_matches options? Why do both exist?

File: requirements.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 Babel==2.3.4
+chardet==3.0.4
```
Comment (d-fence) [inline requirements.txt]: @xmo-odoo looks good to me from the packaging point of view.
The latest ubuntu packages for p2 and p3 are at version 3.0.4 as required.
Fedora 27 and 28 are at the same version.

**Comment by msh-odoo** [comment]
LGTM

**Comment by jat-odoo** [comment]
@xmo-odoo When user import file successfully then `ImportMapping` will store its labels with corresponding fields for that model. Here is nothing related to translate field label (This is another point https://github.com/odoo/odoo/pull/24854/commits/e295ee544ebab5e294535fcf9e448ae9c58d7c50).

**Comment by xmo-odoo** [comment]
> When user import file successfully then ImportMapping will store its labels with corresponding fields for that model.

OK so the point is to store mappings between import column names and actual fields such that if the user had to explicitly provide a mapping, next time they do a similar import with a non-auto-matched (or incorrectly matched) header we remember their previous configuration and they don't have to input it again?

File: addons/base_import/models/base_import.py [hunk 4]
```diff
@@ -453,10 +473,19 @@ def _match_headers(self, rows, fields, options):
...
-        return headers, {
-            index: [field['name'] for field in self._match_header(header, fields, options)] or None
-            for index, header in enumerate(headers)
-        }
+        matches = {}
+        mapping_fields = self.env['base_import.mapping'].search_read([('res_model', '=', self.res_model)], ['column_name', 'field_name'])
+        for index, header in enumerate(headers):
+            match_fields = []
+            for mapping_field in mapping_fields:
+                if mapping_field['column_name'] == header.lower():
```
Comment (xmo-odoo) [inline addons/base_import/models/base_import.py]: mapping_fields should be a mapping of `{column_name: field_name}` so there is no need to repeatedly iterate over it.

The variables should also be renamed & the model should be documented as a mapping of previous header:field selections or some such.

File: addons/base_import/models/base_import.py [hunk 5]
```diff
@@ -453,10 +473,19 @@ def _match_headers(self, rows, fields, options):
...
+                    match_fields = mapping_field['field_name'].split('/') if mapping_field['field_name'] else []
+            if not match_fields:
+                for field in self._match_header(header, fields, options):
+                    if field['name']:
+                        match_fields.append(field['name'])
```
Comment (xmo-odoo) [inline addons/base_import/models/base_import.py]: `_match_headers` seem to assume all fields will have a name, so this seems completely unnecessary, just use e.g.

    match_fields = [f['name'] for f in self._match_header(…)]

File: addons/base_import/static/src/js/import_action.js [hunk 6]
```diff
@@ -591,7 +592,10 @@ StateMachine.create({
         { name: 'import', from: ['preview_success', 'results'], to: 'importing' },
         { name: 'import_succeeded', from: 'importing', to: 'imported'},
         { name: 'import_failed', from: 'importing', to: 'results' }
-    ]
+    ],
+    error: function(eventName, from, to, args, errorCode, errorMessage, originalException) {
+        console.error('event occur from ' + eventName + ': ' + errorMessage);
+    },
```
Comment (xmo-odoo) [inline addons/base_import/static/src/js/import_action.js]: No. The entire point of using a state machine is to have clear and complete transitions, there is no good reason why state transition errors should happen, if they do there's a bug in the code (a case not handled or some such) and it certainly shouldn't be shoved under the rug and hidden away.

File: addons/web/static/src/js/widgets/data_export.js [hunk 7]
```diff
@@ -56,10 +56,19 @@ var DataExport = Dialog.extend({
...
-        'dblclick .o_export_tree_item': function(e) {
+        'dblclick .o_export_tree_item:not(.haschild)': function(e) {
+            var self = this;
             var $elem = $(e.currentTarget);
             $elem.removeClass('o_selected');
             this.add_field($elem.data('id'), $elem.find('.o_tree_column').first().text());
+            // Add its parents to export
+            var $parents = $elem.parents('.o_export_tree_item');
+            if ($parents.length) {
+                _.each($parents, function (el) {
```
Comment (xmo-odoo) [inline addons/web/static/src/js/widgets/data_export.js]: Why is exporting a sub-field automatically adding a bunch of irrelevant garbage to the export? In an awful order to boot, and with unclear behaviour: the parent field was made non-selectable, but it's now added and blocks the addition of the semantically identical but nominally separate External ID.

File: addons/web/static/src/js/widgets/data_export.js [hunk 8]
```diff
@@ -56,10 +56,19 @@ var DataExport = Dialog.extend({
                 }
             }
         },
-        'dblclick .o_export_tree_item': function(e) {
+        'dblclick .o_export_tree_item:not(.haschild)': function(e) {
```
Comment (xmo-odoo) [inline addons/web/static/src/js/widgets/data_export.js]: Why not just remove the o_export_tree_item from rows with children?

File: addons/web/static/src/js/widgets/data_export.js [hunk 9]
```diff
@@ -14,7 +14,7 @@ var _t = core._t;
 var DataExport = Dialog.extend({
     template: 'ExportDialog',
     events: {
-        'click .o_expand_parent': function(e) {
+        'click .o_expand': function(e) {
```
Comment (xmo-odoo) [inline addons/web/static/src/js/widgets/data_export.js]: No. Fix the toggling of icons instead, there is no reason to create a redundant-but-slightly-different class because the icon toggling doesn't work OOTB.

File: addons/web/static/src/xml/base.xml [hunk 10]
```diff
@@ -1328,15 +1328,20 @@
...
+    <t t-foreach="fields" t-as="field">
+        <t t-set="has_child" t-value="field.children &amp;&amp; (field.id).split('/').length != 3" />
+        <div t-att-data-id="field.id" t-attf-class="o_export_tree_item #{has_child ? 'haschild' : ''}" tabindex="-1" role="treeitem"> <!-- tabindex make the div focusable -->
+            <t t-if="has_child">
+                <span class="o_expand o_expand_parent fa fa-chevron-right" role="img" aria-label="Expand parents" title="Expand parents"/>
+                <span class="o_tree_column o_expand" t-att-title="debug and field.id or None"><t t-esc="field.string"/></span>
+            </t>
+            <t t-else="">
+                <span class="o_tree_column" t-att-title="debug and field.id or None"><t t-esc="field.string"/></span>
+            </t>
```
Comment (xmo-odoo) [inline addons/web/static/src/xml/base.xml]: Why duplicate the thing entirely rather than add just the icon and relevant class conditionally?

File: addons/web/static/src/js/core/ajax.js [hunk 11]
```diff
@@ -310,18 +310,29 @@ function get_file(options) {
...
-                    var body = this.contentDocument.body;
-                    var nodes = body.children.length === 0 ? body.childNodes : body.children;
                     var errorParams = {};
-
-                    try { // Case of a serialized Odoo Exception: It is Json Parsable
-                        var node = nodes[1] || nodes[0];
-                        errorParams = JSON.parse(node.textContent);
-                    } catch (e) { // Arbitrary uncaught python side exception
+                    var head = this.contentDocument.head;
+                    if (head.childNodes.length && head.childNodes[0].textContent === '504 Gateway Time-out') { // Check is request timeout(esp. if you set a low proxy_read_timeout in nginx)?
```
Comment (xmo-odoo) [inline addons/web/static/src/js/core/ajax.js]: That seems extremely and unnecessarily brittle.

@ged-odoo do you know if modern web APIs would allow us to get rid of the iframe hack to download files? Or if there's a clean/proper way to detect an error response of some sort (e.g. 504) aside from parsing the page content and hoping it has the right bits?

File: addons/web/static/src/js/widgets/data_export.js [hunk 12]
```diff
@@ -496,7 +496,19 @@ var DataExport = Dialog.extend({
                 import_compat: !!this.$import_compat_radios.filter(':checked').val(),
             })},
             complete: framework.unblockUI,
-            error: crash_manager.rpc_error.bind(crash_manager),
+            error: function (error) {
+                if (error.code === 504 && error.message === '504 Gateway Time-out') {
```
Comment (xmo-odoo) [inline addons/web/static/src/js/widgets/data_export.js]: Why not handle this in the crash manager directly, and why change the format of the error object completely when there's still a hard dependency on the message itself?

File: addons/web/static/src/xml/base.xml (hunk 10 above)
Comment (jat-odoo) [reply addons/web/static/src/xml/base.xml]: @xmo-odoo Duplication is for readability purpose, Otherwise there will be many conditions and it will be look weird.

File: addons/web/static/src/js/widgets/data_export.js (hunk 12 above)
Comment (jat-odoo) [reply addons/web/static/src/js/widgets/data_export.js]: @xmo-odoo  We already handled this in `crash_manager` but that is generic error message. We have to show specific error message when timeout happen during export large amount of data.

File: addons/web/static/src/js/widgets/data_export.js (hunk 7 above)
Comment (jat-odoo) [reply addons/web/static/src/js/widgets/data_export.js]: If user add child field to export then we automatically add its parent field's `External ID` to export list

File: addons/web/static/src/js/widgets/data_export.js (hunk 8 above)
Comment (jat-odoo) [reply addons/web/static/src/js/widgets/data_export.js]: We can't remove `.o_export_tree_item` class because it's already used by its children at many place.

File: addons/web/static/src/js/widgets/data_export.js (hunk 9 above)
Comment (jat-odoo) [reply addons/web/static/src/js/widgets/data_export.js]: We create new class `.o_expand` because if we use class `.o_expand_parent` for both (icon and parent row) then style(css) of `.o_expand_parent` will be applied to parent row too.

File: addons/base_import/static/src/js/import_action.js [hunk 13]
```diff
@@ -111,9 +109,22 @@ var DataImport = AbstractAction.extend(ControlPanelMixin, {
         action.display_name = _t('Import a File'); // Displayed in the breadcrumbs
         this.do_not_change_match = false;
     },
+    /**
+     * @override
+     */
+    willStart: function () {
+        var self = this;
+        var def = this._rpc({
```
Comment (ged-odoo) [inline addons/base_import/static/src/js/import_action.js]: i think you can just return here, and not call this._super

File: addons/web/static/src/js/core/ajax.js (hunk 11 above)
Comment (ged-odoo) [reply addons/web/static/src/js/core/ajax.js]: @xmo-odoo no idea. Could not find much in a quick search. This may be interesting: https://stackoverflow.com/a/36082038

but it uses the fetch API, which is not supported by IE11.  Maybe we could use it anyway, and have a fallback if it does not work, with this code here

Comment (xmo-odoo) [reply addons/web/static/src/js/core/ajax.js]: @ged-odoo ah right I'd completely forgotten about the Fetch API. Might be a good idea to implement get_file on Fetch with an iframe fallback, that way we can cleanly handle issues in the fetch-based version and people using older browser have the less-useful but still-mostly-working version.

Comment (xmo-odoo) [reply addons/web/static/src/js/core/ajax.js]: mmm it looks like Fetch might not even be necessary (though would be nice for better readability): xhr was extended to support [form data](https://developer.mozilla.org/en-US/docs/Web/API/FormData) as well as [blob responses](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType). From there on we can just use download.js or lift the relevant bits from it to actually download the file.

File: addons/base_import/static/src/js/import_action.js (hunk 13 above)
Comment (jat-odoo) [reply addons/base_import/static/src/js/import_action.js]: @ged-odoo Done.

File: addons/web/static/src/js/core/ajax.js (hunk 11 above)
Comment (jat-odoo) [reply addons/web/static/src/js/core/ajax.js]: @xmo-odoo @ged-odoo What we have to do with this point? Final conclusion?

File: addons/web/static/src/js/widgets/data_export.js (hunk 12 above)
Comment (xmo-odoo) [reply addons/web/static/src/js/widgets/data_export.js]: @jat-odoo I've rebuilt the entire get_file function & handled this case in what I think is a cleaner way in #25731 using updated/modern browser APIs obviating the need for iframes. Assuming @ged-odoo is OK with merging that I'll just remove this change from here when merging it.

File: addons/web/static/src/js/core/ajax.js (hunk 11 above)
Comment (xmo-odoo) [reply addons/web/static/src/js/core/ajax.js]: @jat-odoo I've rebuilt the entire get_file function & handled this case in what I think is a cleaner way in #25731 using updated/modern browser APIs obviating the need for iframes. Assuming @ged-odoo is OK with merging that I'll just remove this change from here when merging it.

Comment (jat-odoo) [reply addons/web/static/src/js/core/ajax.js]: :+1:, Thanx for informing.
